Allow to specify web proxy

ChristophAlbert opened the pull request:
[Issue 123](https://github.com/aspnet/SignalR-Client-Cpp/issues/123)

Introduced class signalr_client_config, which contains `web::http::client::http_client_config` and `web::websockets::client::websocket_client_config`.

`hub_connection_tests.mirror_header` was added to test that setting headers still works as expected. No automatic tests were added for using proxies and credentials. It was tested manually that setting a proxy (with basic auth proxy credentials) works, and that `signalr_client_config::http_client_config::set_nativehandle_options` works on Win32.

Commit:
123 Allow to specify web proxy

Commit:
Issue 123 Allow to specify web proxy

Added class signalr_client_config which holds http_client_config and websocket_client_config

dnfclas commented:
Hi __@ChristophAlbert__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

moozzyk commented:
@ChristophAlbert - thanks for you PR. You will need however to sign the CLA before I can move forward.

ChristophAlbert commented:
@moozzyk - I'll try to get this done asap, but since the holiday season in Germany has already started, this might take a little while...

moozzyk commented:
@ChristophAlbert - that's fine. We will resume after the holiday season then. 

Fröhliche Weihnachten!

dnfclas commented:
__@ChristophAlbert__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

ChristophAlbert commented:
It went faster than expected. Frohe Weihnachten ;)

dnfclas commented:
Hi __@ChristophAlbert__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, DNFBOT;

moozzyk commented on include/signalrclient/connection.h:
```diff
@@ -40,6 +41,7 @@ namespace signalr
         SIGNALRCLIENT_API void __cdecl set_disconnected(const std::function<void __cdecl()>& disconnected_callback);
 
         SIGNALRCLIENT_API void __cdecl set_headers(const std::unordered_map<utility::string_t, utility::string_t>& headers);
+        SIGNALRCLIENT_API void __cdecl set_client_config(signalr_client_config config);
```
pass as const reference

moozzyk commented on include/signalrclient/hub_connection.h:
```diff
@@ -43,6 +44,7 @@ namespace signalr
         SIGNALRCLIENT_API void __cdecl set_disconnected(const std::function<void __cdecl()>& disconnected_callback);
 
         SIGNALRCLIENT_API void __cdecl set_headers(const std::unordered_map<utility::string_t, utility::string_t>& headers);
+        SIGNALRCLIENT_API void __cdecl set_client_config(signalr_client_config config);
```
pass as const reference

moozzyk commented on include/signalrclient/signalr_client_config.h:
```diff
@@ -0,0 +1,59 @@
+#include <memory>
+#include "_exports.h"
+
+namespace web
```
why not just include headers?

moozzyk commented on include/signalrclient/signalr_client_config.h:
```diff
@@ -0,0 +1,59 @@
+{
+    //If you need access to the underlying configuration objects, you need to include
+    //cpprest/http_client.h and/or cpprest/ws_client.h
+    class signalr_client_config
```
I am wondering if it wouldn't be better if we rolled headers into this class instead of passing them separately all over the place in addition to config.

moozzyk replied:
and because we already expose cpprestsdk types it would be http_headers instead of the unordered_map

ChristophAlbert replied:
Yes, that would surely make for a nicer API (and implementation). Would you keep `(hub_)connection::set_headers(const std::unordered_map<utility::string_t, utility::string_t>& headers)`, or do you think it can be removed?

moozzyk replied:
I would remove `set_headers`

moozzyk commented on include/signalrclient/signalr_client_config.h:
```diff
@@ -0,0 +1,59 @@
+        // using connection::set_headers or hub_connection::set_headers
+        SIGNALRCLIENT_API void set_credentials(const web::credentials &cred);
+
+        SIGNALRCLIENT_API web::http::client::http_client_config& http_client_config();
```
Returning a reference is problematic. I would prefer returning just the value/copy - I don't think these structs are big so the overhead will be minimal.

moozzyk replied:
If you return a reference just to enable modifying the config - why not just create an API that takes http_client_config (and websocket_client_config)?

ChristophAlbert replied:
The copy overhead will be small; I was more concerned about the compilation times. 
But sure, including the Casablanca headers has a number of advantages - the move operations and unique_ptrs can be removed, the setters/getters be simplified and the client can use the full API without having to include anything in addition.
I'll change the code accordingly.

moozzyk commented on include/signalrclient/signalr_client_config.h:
```diff
@@ -0,0 +1,59 @@
+        SIGNALRCLIENT_API void set_credentials(const web::credentials &cred);
+
+        SIGNALRCLIENT_API web::http::client::http_client_config& http_client_config();
+        SIGNALRCLIENT_API web::websockets::client::websocket_client_config& websocket_client_config();
```
same as above

moozzyk commented on src/signalrclient/connection_impl.cpp:
```diff
@@ -99,7 +99,8 @@ namespace signalr
             {
                 return request_sender::negotiate(*connection->m_web_request_factory, connection->m_base_url,
-                    connection->m_connection_data, connection->m_query_string, connection->m_headers);
+                    connection->m_connection_data, connection->m_query_string, connection->m_headers,
```
roll headers into the config and pass the entire config

moozzyk commented on src/signalrclient/connection_impl.cpp:
```diff
@@ -775,6 +779,12 @@ namespace signalr
         m_headers = headers;
     }
 
+    void connection_impl::set_client_config(signalr_client_config config)
```
const reference

moozzyk commented on src/signalrclient/connection_impl.h:
```diff
@@ -50,6 +51,7 @@ namespace signalr
         void set_reconnected(const std::function<void()>& reconnected);
         void set_disconnected(const std::function<void()>& disconnected);
         void set_headers(const std::unordered_map<utility::string_t, utility::string_t>& headers);
+        void set_client_config(signalr_client_config config);
```
const reference

moozzyk commented on src/signalrclient/hub_connection.cpp:
```diff
@@ -66,4 +66,9 @@ namespace signalr
         m_pImpl->set_headers(headers);
     }
+
+    void hub_connection::set_client_config(signalr_client_config config)
```
const ref

moozzyk commented on src/signalrclient/hub_connection_impl.cpp:
```diff
@@ -267,6 +267,11 @@ namespace signalr
         m_connection->set_headers(headers);
     }
 
+    void hub_connection_impl::set_client_config(signalr_client_config config)
```
const ref

moozzyk commented on src/signalrclient/hub_connection_impl.h:
```diff
@@ -44,6 +44,7 @@ namespace signalr
         utility::string_t get_connection_token() const;
 
         void set_headers(const std::unordered_map<utility::string_t, utility::string_t>& headers);
+        void set_client_config(signalr_client_config config);
```
const ref

moozzyk commented on src/signalrclient/signalr_client_config.cpp:
```diff
@@ -0,0 +1,69 @@
+
+    signalr_client_config::signalr_client_config()
+        : m_http_client_config{ std::make_unique<web::http::client::http_client_config>() }
+        , m_websocket_client_config{ std::make_unique<web::websockets::client::websocket_client_config>() }
```
nit: coma should go on the previous line

moozzyk commented on src/signalrclient/signalr_client_config.cpp:
```diff
@@ -0,0 +1,69 @@
+
+    signalr_client_config::signalr_client_config(const signalr_client_config& other)
+        : m_http_client_config{ std::make_unique<web::http::client::http_client_config>(*other.m_http_client_config) }
+        , m_websocket_client_config{ std::make_unique<web::websockets::client::websocket_client_config>(*other.m_websocket_client_config)}
```
ditto

moozzyk commented on src/signalrclient/signalr_client_config.cpp:
```diff
@@ -0,0 +1,69 @@
+        m_websocket_client_config = std::make_unique<web::websockets::client::websocket_client_config>(*other.m_websocket_client_config);
+        return *this;
+    }
+
```
nit: remove one empty line

moozzyk commented on include/signalrclient/signalr_client_config.h:
```diff
@@ -0,0 +1,59 @@
+        SIGNALRCLIENT_API web::websockets::client::websocket_client_config& websocket_client_config();
+
+    private:
+        std::unique_ptr<web::http::client::http_client_config> m_http_client_config;
```
Is there any specific reasons why these need to be pointers?

ChristophAlbert replied:
`cpprest/http_client.h` and `cpprest/ws_client.h` are quite large and complicated, and include a lot of other stuff. As it might not always be necessary to have very fine control over the options, I think it makes sense to forward-declare the classes here, in order to not slow down the compilation times too much.

Alternative would be a PIMPL, but then we'd exchange the two pointers for one pointer to the implementation...

moozzyk replied:
I would think that with precompiled headers the impact on compilation time would be minimal...

moozzyk commented on src/signalrclient/web_request.cpp:
```diff
@@ -7,8 +7,9 @@
-    web_request::web_request(const web::uri &url)
+    web_request::web_request(const web::uri &url, web::http::client::http_client_config client_config)
         : m_url(url)
+        , m_client_config(std::move(client_config))
```
nit: move to the previous line

moozzyk commented on src/signalrclient/web_request.cpp:
```diff
@@ -7,8 +7,9 @@
 namespace signalr
 {
-    web_request::web_request(const web::uri &url)
+    web_request::web_request(const web::uri &url, web::http::client::http_client_config client_config)
```
const ref

moozzyk replied:
This is quite inconsistent here. Headers are set using the `set_headers` method. Config is passed to the ctor. And we also have the `set_user_agent` method. 
Again, I think headers should be part of the config. The config could be passed to the ctor or to the `get_response()` but it is really used only in the `get_response()` method so probably it can be passed there without having to store it at all

moozzyk commented on src/signalrclient/web_request_factory.cpp:
```diff
@@ -5,11 +5,13 @@
 namespace signalr
 {
-    std::unique_ptr<web_request> web_request_factory::create_web_request(const web::uri &url)
+    std::unique_ptr<web_request> web_request_factory::create_web_request(const web::uri &url, const web::http::client::http_client_config& client_config)
```
const ref

moozzyk commented on test/signalrclient-testhost/Connections/HubConnection.cs:
```diff
@@ -80,5 +80,11 @@ public class Address
+
+        public string MirrorHeader()
+        {
+          var mirrorValue = Context.Request.Headers["x-mirror"];
```
nit: spacing

moozzyk commented on src/signalrclient/signalr_client_config.cpp:
```diff
@@ -0,0 +1,69 @@
+        m_websocket_client_config->set_proxy(proxy);
+    }
+
+    void signalr_client_config::set_credentials(const web::credentials &cred)
```
nit: credentials instead of cred

moozzyk commented on include/signalrclient/signalr_client_config.h:
```diff
@@ -0,0 +1,59 @@
+        // For example, Basic Authentication fails under Win32.
+        // As a workaround, you can set the required headers directly by
+        // using connection::set_headers or hub_connection::set_headers
+        SIGNALRCLIENT_API void set_credentials(const web::credentials &cred);
```
nit: credentials instead of cred

moozzyk reviewed (changes requested):
I think we should roll headers into the signalr config and pass the entire config around by const reference. Only the final consumer should 'slice' the config (i.e. pick the part they need).

ChristophAlbert committed:
Issue 123 Allow to specify web proxy

- Changed signalr_client_config such that it has the config objects as members. For that, included casablanca headers directly
- Moved headers from connection_impl to signalr_client_config. Changed type of headers to web::http::http_headers
- Removed (hub_)connection::set_headers and replaced it with set_client_config
- Configs are now passed around as const refs
- Removed web_request::set_headers. Instead, web_request::get_response now takes a signalr_client_config as a parameter
- Adapted tests accordingly

ChristophAlbert committed:
Fixed comment of signalr_client_config::set_credentials regarding authorization

moozzyk commented on src/signalrclient/connection_impl.cpp:
```diff
@@ -769,10 +772,10 @@ namespace signalr
     {
-        ensure_disconnected(_XPLATSTR("cannot set headers when the connection is not in the disconnected state. "));
-        m_headers = headers;
+        ensure_disconnected(U("cannot set client config when the connection is not in the disconnected state. "));
```
Generally the product code uses the `_XPLATSTR` macro and not `U`, so replacing.

moozzyk commented on src/signalrclient/http_sender.cpp:
```diff
@@ -11,15 +10,15 @@ namespace signalr
             request->set_user_agent(USER_AGENT);
 
-            return request->get_response().then([](web_response response)
+            return request->get_response(signalr_client_config).then([](web_response response)
```
It should be possible to set the config separately. In case of sending many related http requests (e.g. longpolling) it is important to reuse the same instance of the http client and setting the config separately makes it easier. (Currently not a problem though because this client does not support long polling).

moozzyk commented on src/signalrclient/transport_factory.cpp:
```diff
@@ -8,13 +8,13 @@
         if (transport_type == signalr::transport_type::websockets)
         {
-            return websocket_transport::create([headers](){ return std::make_shared<default_websocket_client>(headers); },
+            return websocket_transport::create([&signalr_client_config](){ return std::make_shared<default_websocket_client>(signalr_client_config); },
```
I think it is safer for this to be captured by value.

moozzyk commented on test/signalrclient-e2e-tests/hub_connection_tests.cpp:
```diff
@@ -473,3 +474,39 @@ TEST(hub_connection_tests, connection_token_start_stop_start_reconnect)
+    headers[U("x-mirror")] = U("MirrorThat");
+    signalr_client_config.set_http_headers(headers);
+
+    ASSERT_THROW(hub_conn->set_client_config(signalr_client_config), signalr::signalr_exception);
```
This is being tested elsewhere so, removing.

moozzyk commented on test/signalrclienttests/test_transport_factory.h:
```diff
@@ -14,9 +14,10 @@ class test_transport_factory : public transport_factory
         std::function<void(const utility::string_t&)> process_message_callback,
-        std::function<void(const std::exception&)> error_callback) override;
+        std::function<void(const std::exception&)> error_callback,
+        const signalr_client_config& signalr_client_config = signalr_client_config{})
```
I tried to follow a convention where callbacks are the last parameters so I will change the order.

moozzyk commented on test/signalrclient-e2e-tests/hub_connection_tests.cpp:
```diff
@@ -473,3 +474,39 @@ TEST(hub_connection_tests, connection_token_start_stop_start_reconnect)
+    hub_conn->set_client_config(signalr_client_config);
+
+    {
+        auto test = hub_conn->start().then([&hub_proxy]()
```
the name of the variable needs to be more descriptive

moozzyk reviewed (approved):
Looks solid. I have a few comments. I created a follow up PR to address them https://github.com/aspnet/SignalR-Client-Cpp/pull/158. Let me know what you think.

moozzyk commented:
I squashed and rebased your PR and merged as 3bc6b067c83d5fda9abbf3909a0311889dde6277. New packages containing your changes are ready on the https://dotnet.myget.org/F/aspnetcore-ci-dev/api/v3/index.json feed (gallery: https://dotnet.myget.org/gallery/aspnetcore-ci-dev). 

Thanks a lot for your contribution!
